### PR TITLE
Testing practice

### PR DESCRIPTION
**Pull request by imhilla**
Hello TSE,
Please review our testing practice project for the JavaScript section. We added functions and respective test files for each.
Please provide good feedback as that will be appreciated.
We added:-
- Calculator function
- Capitalize function
-  Reverse function
- Tests
- Lastly, cipher function

Hillary

### CODE REVIEW

**Review by mikenath223 — CHANGES_REQUESTED** [review]
### Hi Team. 👋

**Nice work on this.**

- Please add a test on ceasarCipher that checks for keeping the same case and for testing that punctuation is returned as expected.

Kindly fix these issues and request another review👍
Feel free to contact me on slack(michgolden) if you need further clarification.

HAPPY CODING👨‍💻

File: cipher.test.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+});
+
+test('ciphers strings', () => {
+  expect(cipher('abcdefghijklmnopqrstuvwxyz', 1)).toBe('bcdefghijklmnopqrstuvwxyza');
+  expect(cipher('abcde', 5)).toBe('fghij');
+});
+
+test('should throw an error', () => {
+  expect(cipher('wffeg.g', 3)).toBeTruthy();
+});
```
Comment (mikenath223) [inline cipher.test.js]: - [ ] Include tests for punctuation and capitalize.

**Review by mikenath223 — CHANGES_REQUESTED** [review]
### Hi again Team. 👋

Nice work so far in fixing some issues raised.

- [ ] Please note that in testing ceasar_cipher if a char is capitalized the encrypted returned char should also be capitalized. Kindly fix this.

- [ ] - IMPORTANT - Your linter isn't properly configured. There is a recent update to `.eslintrc.json` and `.github/workflows/linters.yml`. Please refer to [this link](https://github.com/microverseinc/linters-config/tree/master/javascript) and update yours accordingly.

Kindly fix these issues and request another review👍
Feel free to contact me on slack(michgolden) if you need further clarification.

HAPPY CODING👨‍💻

File: cipher.test.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+test('ciphers strings', () => {
+  expect(cipher('abcdefghijklmnopqrstuvwxyz', 1)).toBe('bcdefghijklmnopqrstuvwxyza');
+  expect(cipher('abcde', 5)).toBe('fghij');
+});
+
+test('keeps the same case regardless of input', () => {
+  expect(cipher('abcDefghiJklmnOpqrstUvwxyz', 1)).toBe('bcdefghijklmnopqrstuvwxyza');
+});
```
Comment (mikenath223) [inline cipher.test.js]: - [ ] The expected return should rather be "bcDefghijKlmnoPqrstuVwxyza".

File: arrayAnalysis.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  return total / arrayValue.length;
+};
+
+function analyze(arrayValue) {
+  const maxValue = Math.max(...arrayValue);
+  const minValue = Math.min(...arrayValue);
+  const arrayLength = arrayValue.length;
+  const averageValue = averagefunction(arrayValue);
+  return {
+    average: averageValue, min: minValue, max: maxValue, length: arrayLength,
```
Comment (uryelah) [inline arrayAnalysis.js:15]: - [ ] [OPTIONAL] You could name each method already with the name you wish to return them with and return the object like:
`return {average, min, max, length}`

Comment (uryelah) [inline arrayAnalysis.js:15]: [OPTIONAL] You could name each method already with the name you wish to return them with and return the object like:
return {average, min, max, length}

File: arrayAnalysis.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+const averagefunction = (arrayValue) => {
+  let total = 0;
+  arrayValue.forEach((value) => {
+    total += value;
+  });
+  return total / arrayValue.length;
```
Comment (uryelah) [inline arrayAnalysis.js:6]: - [ ] [OPTIONAL] What happens if the length is 0 though? You could handle that edge case here already to make the testing easier.

File: arrayAnalysis.test.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import analyze from './arrayAnalysis';
+
+test('Analyze the Max Value in array', () => {
+  expect(analyze([1, 5, 6]).max).toBe(6);
```
Comment (uryelah) [inline arrayAnalysis.test.js:4]: - [ ] [OPTIONAL] For your next tests make sure to always test the methods with more than one test case. If you only test it with one specific array it's not possible to be completely sure the methods will also work properly with other arrays, especially ones with edge cases like an empty array for example.
[This article about testing](https://outline.com/https://itnext.io/test-driven-development-101-with-jest-a6e666d090a5) could be useful to guide your future approach to tests. Pay attention specially to the parts to do with 'triangulation'.

File: calculator.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  add: (a, b) => {
+    const x = a + b;
+    return x;
+  },
+  subtract: (a, b) => {
+    const x = a - b;
+    return x;
+  },
+  divide: (a, b) => {
+    const x = a / b;
```
Comment (uryelah) [inline calculator.js:11]: - [ ] [OPTIONAL] Again, think about the edge cases. What would happen if b was 0?

File: capitalize.js [hunk 7]
```diff
@@ -0,0 +1,2 @@
+const capitalize = (string) => string.charAt(0).toUpperCase() + string.slice(1);
```
Comment (uryelah) [inline capitalize.js:1]: - [ ] [OPTIONAL] One neat fact about strings in JS is that you can get a character from it using the bracket notation like with arrays:

`string[0]` would have the same letter as `string.charAt(0)`

That could make your code simpler.

File: cipher.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+const isLowerCase = stringI => (/[a-z]/.test(stringI));
+
+const cipher = (string, key) => {
+  if ((regTest(string)) === true) {
+    return true;
+  }
+  const stripped = string.split('');
+  const empty = [];
+  for (let i = 0; i < stripped.length; i += 1) {
```
Comment (uryelah) [inline cipher.js:12]: - [ ] [OPTIONA] Another fun fact on strings in JS is that you can use a `for of` loop to iterate through a string without having to convert it into an array.

[Check this MDN article on that feature.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of)

File: cipher.test.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  expect(cipher('abcdefghijklmnopqrstuvwxyz', 1)).toBe('bcdefghijklmnopqrstuvwxyza');
+  expect(cipher('abcde', 5)).toBe('fghij');
+});
+
+test('keeps the same case regardless of input', () => {
+  expect(cipher('abcDefghiJklmnOpqrstUvwxyz', 1)).toBe('bcdEfghijKlmnoPqrstuVwxyza');
+});
+
+test('should throw an error', () => {
+  expect(cipher('wffeg.g', 3)).toBeTruthy();
```
Comment (uryelah) [inline cipher.test.js:18]: - [ ] [OPTIONAL] How would you write your Cypher method in a way that it works with punctuation?
Ex: 
```
cipher('ab.a', 1)
=> "bc.b"
```
Think about how you could keep the position/index of the special characters in the final string.

File: README.md [hunk 10]
```diff
@@ -1 +1,80 @@
...
+- HTML
+- Git
+- CSS
+- JavaScript
+- Shields.io
+- Webpack
+- npm
+- command-line tools such as git.
+
+## Test with
```
Comment (uryelah) [inline README.md:41]: - [ ] This is not enough to run the tests locally, make sure to add instructions on downloading the project and installing the packages with npm before it.
Be as detailed and specific as possible in the README instructions, avoid assuming your user knows how to run the code/tests.

**Review by uryelah — APPROVED** [review]
Hi **Hillary Odhiambo & Muhammad Umer**, I'm Sarah your reviewer for this PR 👩‍🚀 

Good job starting your journay into testing JS!

While there were some issues with the project, since they were not pointed out at the first review I'm marking them as optional but I really reccomend you to fix them before mergin this PR.

The main issues would be not separating the files between logic and test folders, the README lacking the setup instructions, and not completelly testing the ponctuation for the Caesar Cypher(though this one was left vague at the Odin project instrucitons).

- [ ] [OPTIONAL] Each module/class/factory should have its own file, and each file should have a corresponding test file. For example, there should be a calculator.test.js for a calculator.js file. 

Organizing your code in the above way will make your repo more professional and is overall more makes for a more manageable architecture as your project grows.

-------

## **Your project is approved! 🎊 🥅 💛**

**You may close this Pull Request and merge the changes**


---------

    
    
**Keep being awesome devs and good luck with your next project!**

---------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)
